### PR TITLE
Ho node manager types

### PR DESCRIPTION
**Pull request by ajreynol**
This introduces the notion of first-class type.  Function type is first-class type if options::ufHo() is enabled.  

Updates the assertions in NodeManager in type construction.   Also opens issue #1202 regarding equalities between terms whose types are not first-class.

### CODE REVIEW

**Comment by ajreynol** [comment]
This commit is 9/10 planned commits from the ho-smt branch: https://github.com/CVC4/CVC4/tree/ho-smt

File: src/expr/type_node.h [hunk 1]
```diff
@@ -429,6 +429,16 @@ class TypeNode {
    */
   bool isInterpretedFinite() const;
 
+
+  /**
+   * Is this a first-class type?
+   * First-class types are types for which we handle equalities between terms
```
Comment (barrettcw) [inline src/expr/type_node.h]: This wording is a bit confusing.  Is the following rewording correct?

First-class types have the following two properties:
1) we can reason about equalities between terms of that type; and
2) they are allowed to be used to instantiate the parameters of parametric theories (e.g. the index or element types in arrays)

File: src/expr/type_node.cpp [hunk 2]
```diff
@@ -86,6 +87,16 @@ bool TypeNode::isInterpretedFinite() const {
   }
 }
 
+bool TypeNode::isFirstClass() const {
+  return ( getKind() != kind::FUNCTION_TYPE || options::ufHo() ) &&
```
Comment (barrettcw) [inline src/expr/type_node.cpp]: This seems a bit fragile - how do we ensure that this function stays correct?  Maybe something worth discussing or opening an issue for - not necessarily to be addressed by this PR.

Comment (ajreynol) [reply src/expr/type_node.cpp]: Sure, I suppose we could add this as an attribute in the kinds file.
Notice we have the same problem with maintaining "isSubtypeOf".

Comment (barrettcw) [reply src/expr/type_node.cpp]: Do you mind opening an issue for this?

Comment (ajreynol) [reply src/expr/type_node.cpp]: I've opened #1223.

**Pull request merged**